src/concrete/ml/sklearn/linear_model.py

```diff
@@ -493,102 +556,157 @@ def _fit_encrypted(
+        # the loop)
+        loss_value_moving_average = None
 
+        # Iterate on the training quantized module in the clear
+        for iteration_step in range(max_iter):
+            X_batch_enc_i, y_batch_enc_i = (
+                X_batches_enc[iteration_step],
+                y_batches_enc[iteration_step],
+            )
             # Train the model over one iteration
             inference_start = time.time()
-            weights, bias = self.training_quantized_module.forward(  # type: ignore[assignment]
-                X_batch, y_batch, weights, bias, fhe=fhe
-            )
 
-            if self.verbose:
-                print(
-                    f"Iteration {iteration_step} took {time.time() - inference_start:.4f} seconds."
+            # If the training is done in FHE, execute the underlying FHE circuit directly on the
+            # encrypted values
+            if fhe == "execute":
+                weights_enc, bias_enc = self.training_quantized_module.fhe_circuit.run(
+                    X_batch_enc_i,
+                    y_batch_enc_i,
+                    weights_enc,
+                    bias_enc,
                 )
 
-            # Mypy
-            assert isinstance(weights, numpy.ndarray)
-            assert isinstance(bias, numpy.ndarray)
+            # Else, use the quantized module on the quantized values (works for both quantized
+            # clear and FHE simulation modes). It is important to note that 'quantized_forward'
+            # with 'fhe="execute"' is executing Concrete's 'encrypt_run_decrypt' method, as opposed
+            # to the 'run' method right above. We thus need to separate these cases since values
+            # are already encrypted here.
+            else:
+                weights_enc, bias_enc = self.training_quantized_module.quantized_forward(
+                    X_batch_enc_i, y_batch_enc_i, weights_enc, bias_enc, fhe=fhe
+                )
 
-            # Reshape parameters to fit what scikit-learn expects
-            weights = weights.squeeze(0)
-            bias = bias.squeeze(0)  # pylint: disable=no-member
+            if self.verbose:
+                print(
+                    f"Iteration {iteration_step} took {time.time() - inference_start:.2f} seconds."
+                )
 
-            # If early stopping is enabled, compute the loss and stop the training if it gets under
-            # the given tolerance
+            # If early stopping is enabled, decrypt (if needed) and de-quantize the weight and bias
+            # values. Then, compute the loss and stop the training if it gets under the given
+            # tolerance
             # Additionally, there is no point in computing the following in case of a partial fit,
             # as it only represents a single iteration
             if self.early_stopping and not is_partial_fit:
+                weights_float, bias_float = self._decrypt_dequantize_training_output(
```

Comment:
we keep early possible possible with composition by adding this decrypt/dequant step here (since this is only for development, we believe it's not really an issue to do that)